### ISSUE TITLE
Automatically detect and display file size on dataset page

### ISSUE THREAD

**Issue by philipashlock** [GSA/datagov-wptheme#171]
Unfortunately neither HTTP Content-Length or Content-Type headers tend to be reliable to detect this automatically so some files would need to be completely downloaded to detect actual size and format.

**Comment by ErieMeyer** [comment]
Seems like a lower priority in advance of launch.

**Comment by philipashlock** [comment]
Agreed. I thought this was specified by the POD schema and was going to wait for that to be reflected on the live site before closing. For now, perhaps this should be considered as an Expanded field and requested for inclusion on the POD tracker? https://github.com/project-open-data/project-open-data.github.io/issues

**Comment by MarinaNitze** [comment]
We intentionally removed size from the schema. It's not something people should enter manually, and it's irrelevant for things like APIs. So this ticket should be "automatically detect size of certain file types" if that is important to enough visitors.

**Comment by philipashlock** [comment]
Agreed, though this also applies to 'format' which is part of the POD schema and is often entered incorrectly (See #55). Unfortunately neither HTTP Content-Length or Content-Type headers tend to be reliable to detect this automatically so some files would need to be completely downloaded to detect actual size and format.

**Comment by JeanneHolm** [comment]
+1 on this via Twitter @sherifffruitfly
@usdatagov ok last suggestion quickly: include size in the gloss on the datasets. rowcount/mb/etc

**Comment by rossjones** [comment]
Have you considered using the ckanext-archiver and ckanext-qa combination?  You get better file sizes (because it actually does download the resource) and then QA uses libmagic to determine the actual file type (with reasonably good results). For bonus points you then also get per organisation broken link reports.

**Comment by JeanneHolm** [comment]
@rossjones Great suggestion! I didn't know about this, perhaps @philipashlock or @dwcaraway could weigh in?

**Comment by dwcaraway** [comment]
@rossjones @JeanneHolm I haven't used those ckan extensions, but they look interesting. I think datagov would, more generally, benefit from verifying more of the metadata where possible.

**Comment by JeanneHolm** [comment]
These plug-ins are being tested and look promising.  Work progressing.

**Comment by rossjones** [comment]
It may be worth speaking with @davidread at data.gov.uk to get their optimised version which puts less of a strain on the ckan api.

**Comment by JeanneHolm** [comment]
Thanks! Will do as well.

**Comment by davidread** [comment]
Yes, getting file size and detecting broken links are achieved with:
https://github.com/datagovuk/ckanext-archiver/
It stores the results in a separate table to the main package, so that you don't get lots of revision history for this metadata that can change more frequently. The archiver runs outside the webapp, on a celery queue, triggered by a resource URL change/addition. We also run it on all resources every weekend (using a cron) to see if any of the files have changed size or gone missing (404).

The broken links are reported here: http://data.gov.uk/data/report/broken-links and broken-down by organization. This reporting stuff requires the ckanext-report module and should be pretty easy for you to integrate.

The next step would be to add in ckanext-qa which checks the format of the downloaded files, using 'file' and 'libmagic'. We show 'warnings' like this one: http://data.gov.uk/dataset/live-traffic-information-from-the-highways-agency-road-network/resource/ca9d2d90-277c-4fd4-8d4d-0319182bb715 but we could do more to expose this.

Am happy to help with integration if needs be.

**Comment by JeanneHolm** [comment]
@davidread--thanks, this is excellent! @dialsunny and @kvuppala please reference here.

**Comment by kvuppala** [comment]
this is completed

**Comment by rebeccawilliams** [comment]
Re-opening this issue, because of additional [asks](https://twitter.com/hpcprogrammer/status/572923242446516224) for this feature and new potential paths forward to integrate this feature with other QA checks (see https://github.com/GSA/project-open-data-dashboard/labels/quality%20check)

**Comment by rebeccawilliams** [comment]
Since Data.gov hosts metadata, downloading the data is necessary to see how large the file is. Downloading all of this data is not practical at this time via Data.gov's QA crawls. (see also #459) 

@philipashlock mentioned the Internet Archive may be interested. Noting that here for follow up.

**Comment by davidread** [comment]
The data.gov.uk archiver has a cut-off at 50Mb. If a file's headers say it is bigger than that, or if the download starts to exceed that, then it cancels the download. With that cut-off, the total data we store is currently only 60GB, and if we removed the cut-off it is about twice that. A disk to hold that is under $100, so I don't imagine that size should be a concern to data.gov.
